### PR TITLE
Feature/linuxnogtk

### PR DESCRIPTION
**Pull request by andydotxyz**
Adding support for Linux without GTK or CGo using the dbus protocol.

TODO:
- [x] Support for showing icon
- [x] Move to a cleaner code/XML/schema definition
- [x] Add menu display

Looking for feedback on approach or impl, needs lots of testing as this is a whole new approach compared to upstream

### CODE REVIEW

File: systray_linux.go [hunk 1]
```diff
@@ -18,3 +62,268 @@ func (item *MenuItem) SetIcon(iconBytes []byte) {
...
+}
+
+func quit() {
+	close(quitChan)
+}
+
+func nativeStart() {
+	systrayReady()
+
+	intro := introspect.Introspectable(`<node>
```
Comment (Bluebugs) [inline systray_linux.go:106]: Wouldn't it be possible to use fyne bundle to store the file in git as the xml file directly?

Comment (Bluebugs) [reply systray_linux.go:106]: Additionally have you looked at https://github.com/amenzhinsky/dbus-codegen-go ? I feel like a codegen dbus would be nicer to use and seems less brittle.

File: systray_linux.go [hunk 2]
```diff
@@ -18,3 +62,268 @@ func (item *MenuItem) SetIcon(iconBytes []byte) {
...
+	p, _, err := image.Decode(bytes.NewReader(iconData))
+	if err != nil {
+		log.Printf("Failed to read icon format %v", err)
+		return PX{}
+	}
+
+	w, h := p.Bounds().Dx(), p.Bounds().Dy()
+	data2 := make([]byte, w*h*4)
+	i := 0
+	for y := 0; y < h; y++ {
```
Comment (Bluebugs) [inline systray_linux.go:267]: This really doesn't look great. I think the right way in Go would be to use the draw package to output to a known RGBA typed image and then return rgba.Pix ( https://pkg.go.dev/image#RGBA ).

Comment (andydotxyz) [reply systray_linux.go:267]: Sadly the output format needs to be ARGB not RGBA and none of the Go standard image formats support this to my knowledge.
Is there some way I can make it look less bad within this constraint?

File: systray_linux.go (hunk 1 above)
Comment (andydotxyz) [reply systray_linux.go:106]: Yes, bundling is one possibility for the solution, however as you comment I am not sure that working from XML is the best end solution.
The hint for dbus-codegen looks like just the thing that could tidy this up, thanks.

File: systray_linux.go (hunk 2 above)
Comment (Bluebugs) [reply systray_linux.go:267]: Arg, I didn't realize that. Quite annoying that Go doesn't come with the infrastructure to do ARGB nicely ( https://github.com/golang/go/issues/12966 ). A potential cleaner solution would be to use https://pkg.go.dev/golang.org/x/exp/shiny@v0.0.0-20220218215828-6cf2b201936e/driver/internal/swizzle as pointed by the comment in that issues.

Comment (andydotxyz) [reply systray_linux.go:267]: That is an alternative. However we do have to decode the image first (to get the pixel data and the X/Y size) so is the code above really an overhead?

Comment (andydotxyz) [reply systray_linux.go:273]: Updated to be more readable

File: systray_linux.go (hunk 1 above)
Comment (andydotxyz) [reply systray_linux.go]: Done :)

File: systray_menu_linux.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+				nil,
+			},
+		},
+	}
+}
+
+type menuLayout = struct {
+	V0 int32
+	V1 map[string]dbus.Variant
+	V2 []dbus.Variant
```
Comment (Bluebugs) [inline systray_menu_linux.go:101]: What are V0, V1 and V2 purpose?

Comment (andydotxyz) [reply systray_menu_linux.go]: Good question, added comments to clarify.
I cannot find a way to do better (like with nice variable names) as this is matching auto-generated code...

**Review by Bluebugs — APPROVED** [review]

**Pull request merged**